zstd early abort

rincebrain opened the pull request:
### Motivation and Context
LZ4 is really fast, especially at giving up on incompressible data.

ZSTD is pretty great at compressing things well, but has no such heuristic.

What if...it did?

Well...

Raspberry Pi 4:
![zstd abort Pi WIP](https://user-images.githubusercontent.com/214141/160515419-51de7625-8e4f-4913-a317-fc2e439f54d5.png)

Ryzen 5900X:
![zstd go fast 5900X](https://user-images.githubusercontent.com/214141/159585202-fa6d0f3d-8606-463c-a669-e5bde187d1e0.png)

< 0.3% more space used to go from 2 hours to 15 minutes? Sold.

### Description
As the comment block says:
```
        /*
         * A zstd early abort heuristic.
         *
         * - Zeroth, if this is <= zstd-3, or <= zstd_abort_size (currently 128k),
         *   don't try any of this, just go.
         *   (because experimentally that was a reasonable cutoff for a perf win
         *   with tiny ratio change)
         * - First, we try LZ4 compression, and if it doesn't early abort, we
         *   jump directly to whatever compression level we intended to try.
         * - Second, we try zstd-1 - if that errors out (usually, but not
         *   exclusively, if it would overflow), we give up early.
         *
         *   If it works, instead we go on and compress anyway.
         *
         * Why two passes? LZ4 alone gets you a lot of the way, but on highly
         * compressible data, it was losing up to 8.5% of the compressed
         * savings versus no early abort, and all the zstd-fast levels are
         * worse indications on their own than LZ4, and don't improve the LZ4
         * pass noticably if stacked like this.
         */
```

A graphic of how much space it takes up versus stock zstd-3 if you use one or two passes of this form:
![dang it 2](https://user-images.githubusercontent.com/214141/159585888-6103fcee-a99f-4a06-87f5-13777770d060.png)

I have...so many spreadsheets of data trying different permutations of this on different hardware.

### What's left to do?
Opening the PR to get people's opinions on whether this is worth including, whether they'd like, say, a property on datasets or a module parameter to turn it on and off like I have here, or various other things, before I try to seriously pitch getting it included.

Getting it included would obviously include reordering the commits to, say, not combine some of the code changes and moving lz4 into zcommon, removing the debug tunables, refactoring the zstd compression calls to not be a copy-paste...you get the idea.

Footnote: This trick probably works fine for gzip too.

### How Has This Been Tested?
I've run this through its paces on my Pi 4, my Ryzen 5900X, and my 8700k. (I wanted to try it on Linux/x86, but it turns out OpenZFS is too broken for that to work with unmodified git, let alone any changes.)

I've run through 2 datasets for testing, primarily - the low compression dataset I used, around 45 GB uncompressed, with a compression ratio of 1.05 or so as a baseline, made up of ZIP files of firmware images (copied to two separate datasets with recordsize=128k and recordsize=1M configured), and a high compression dataset (ratio of 2.25 or so), 55 GB or so uncompressed, that's a snapshot of my maildir, again in 128k and 1M format.

I ran b3sum against all the files on the different platforms and compared to the reference, and have so far found no discrepancies, so I don't appear to be secretly mangling bytes.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

gmelikov commented:
Sorry for offtopic - could someone point me at "lz4 early abort mechanism"? All I've found is ZFS's 12.5% threshold, which will be passed through into any compression algo.

rincebrain commented:
Not really offtopic, IMO.

Technically, what I'm proposing is not what LZ4 does - the only thing LZ4 has, AFAIK, other than the normal "oops you hit the buffer size we passed in, give up" is giving up on hunks of data every N if it hasn't found anything compressible in it, not giving up on the whole block (at least, I don't think the constants are large enough for that in most cases...)

In our code, look for references to "SKIPSTRENGTH" - e.g.

https://github.com/openzfs/zfs/blob/6b444cb9711a8c0a089612dc3ac67e45c4a4e108/module/zfs/lz4_zfs.c#L505

Or the more informative prior definition:
https://github.com/openzfs/zfs/blob/6b444cb9711a8c0a089612dc3ac67e45c4a4e108/module/zfs/lz4_zfs.c#L182-L191

In modern LZ4, the definition to look for references to is "LZ4_skipTrigger", albeit much less informative:

https://github.com/lz4/lz4/blob/90d68e37093d815e7ea06b0ee3c168cccffc84b8/lib/lz4.c#L647

(I did a couple of experiments changing that value. I don't recommend it - the changes in perf/compratio are absolutely not worth it.)

adamdmoss commented:
Looks great.  I can't easily think of a downside.  Even the code changes are pretty minimal.

nabijaczleweli commented on module/zstd/zfs_zstd.c:
```diff
@@ -50,6 +50,10 @@
 #include "lib/zstd.h"
 #include "lib/common/zstd_errors.h"
 
+int zstd_lz4_pass = 1;
+int zstd_firstpass_level = ZIO_ZSTD_LEVEL_1;
+unsigned int zstd_abort_size = 131072;
```
`static`?

rincebrain replied:
Thought I wrote a reply to this, but "sure, for any of the ones that remain - I was at least planning to drop the first one unless people feel strongly about it, and probably make the second not tunable."

rincebrain commented on module/zfs/spa_misc.c:
```diff
@@ -244,16 +244,10 @@ static avl_tree_t spa_l2cache_avl;
  */
-int zfs_flags = ~(ZFS_DEBUG_DPRINTF | ZFS_DEBUG_SET_ERROR |
-    ZFS_DEBUG_INDIRECT_REMAP);
-#else
-int zfs_flags = 0;
-#endif
+extern int zfs_flags;
```
That's stolen from #13206 without cleanly pulling in the commit separately - but sure, I'll see if that change flies there and port it appropriately.

beren12 commented:
This looks great btw.

Ornias1993 commented:
Going over the test data, I approve of this.
Currently the go-to standard (default) is either ZSTD-1 or ZSTD-3 (very close tie), with these changes, it allows users to safely go to ZSTD-5 without much negative effect on low-compression workloads.

It's basically the no-comp equivalent of how the current ZSTD implementation already handles cant-alloc cases, so it fits within the design view.

adamdmoss commented:
I've been using this for about a month now with no ill effects, FWIW.  Seems to work great.

adamdmoss commented:
Mind rebasing?  Doesn't apply to current trunk.

rincebrain commented:
There, now with many of the knobs removed.

I'll probably remove the rest shortly, except the "on/off", since I was told that making it a dataset property was unreasonable.

rincebrain commented:
Or the rebase could cause memory corruption on older kernels despite not modifying the allocation codepaths (or even directly calling them outside of the existing function implementations). That's neat. I wonder what on earth is going on.

Time to break out the KASAN I guess.

nabijaczleweli commented on module/zstd/zfs_zstd.c:
```diff
@@ -50,7 +50,14 @@
+static int zstd_cutoff_level = ZIO_ZSTD_LEVEL_3;
+static int zstd_zstd_pass = 1;
+static int zstd_hard_moed = 0;
+static unsigned int zstd_abort_size = (131072);
```
`128 * 1024`?

rincebrain replied:
Sure.

nabijaczleweli commented on module/zstd/zfs_zstd.c:
```diff
@@ -50,7 +50,14 @@
+static int zstd_hard_moed = 0;
+static unsigned int zstd_abort_size = (131072);
+
+kstat_t *zstd_ksp = NULL;
```
I don't see why this needs to not be static anymore?

rincebrain replied:
It doesn't, it's just a change that I overlooked when merging.

rincebrain replied:
Fixed, unless I'm worse at reading than I think I am.

rincebrain commented:
Disabling compressed ARC and trying to compress things has broken handling, resulting in sometimes trying to call `zio_compress_data` [here](https://github.com/openzfs/zfs/blob/7eba3891e9d0c211a1650ecfdfc78c654be055f8/module/zfs/arc.c#L9339-L9344) claiming the input buffer is 128k long when it's actually 512b of data, and we allocated 512b.

I tried to guess how it was intended to work, and it at least passes the test and doesn't complain on my KASAN kernel, but I am not remotely confident that I'm certain what is necessary in all cases. So I may just go back to my first solution of "not if you have compressed ARC enabled" unless someone debugs it if it's not fixed everywhere.

rincebrain commented:
Anyway, the addendum I was planning to post today, before the above excitement debugging, which hopefully covers the remaining concerns people voiced.

So, at the last leadership meeting, it was requested that I showed off how much more expensive it could be in the worst case, if we ran through all these preprocessing passes and still decided to do the compression.

So, I did it, on the various test datasets I was using before. (“alwayson” is the mode in which it unconditionally runs all passes on anything sufficiently large, which is still configured to 128k minimum as before.)

(I'm sorry one or two of the graphs still call the mode for "run all compression passes unconditionally" "hardmoed", which is what I referenced it as in the code and used initially, but I only noticed it now when posting this. So it's the same thing, I just missed updating it in one or two templates, apparently.)

The breakdown of compressibility stats goes:
passignored means it was excluded from any EA processing, either from the record being too small or the compression level requested being too low
lz4pass_allowed means “LZ4 compressed it successfully, we stopped further EA processing and went straight to the heavy compression”
zstdpass_allowed means, similarly, “LZ4 rejected it, but zstd-1 compressed it, so we go on to compress it” - this is the most expensive path, if we do all the compression steps
zstdpass_rejected means “LZ4 and zstd-1 both refused to compress it, so we’re not even gonna try”
The sum of the zstdpass_allowed/rejected is the number of records LZ4 rejected 
An interesting quirk is that if you vary the recordsize, the absolute number of passignored records remains relatively constant in my test datasets, even while the number of records large enough varies drastically.

**5900X, low compression, 128k records:**
![128k lowcomp 5900X](https://user-images.githubusercontent.com/214141/164955763-8cbb1ab0-80b9-42bd-9157-62ef994d8319.png)
**5900X, low comp, 1M records:**
![1M lowcomp 5900X](https://user-images.githubusercontent.com/214141/164955766-01b0f52b-5ce6-4cde-861e-97af96aff590.png)
**5900X, high comp, 128k records:**
![128k highcomp 5900X](https://user-images.githubusercontent.com/214141/164955765-9aa71cfc-7e0b-4ebc-8c18-0940e6de5ce8.png)
**5900X, high comp, 1M records:**
![1M highcomp 5900X](https://user-images.githubusercontent.com/214141/164955764-272ed4bd-f863-41fb-86bc-3ec5396cd4f6.png)

If you’d like to see something a little more interesting, I made an evil dataset which fails to compress at zstd-1 but succeeds at zstd-2 and above, and tested it with baseline, with a zstd-1 early abort pass, with a zstd-2 early abort pass, and with modes that unconditionally ran all steps for each (which don’t differ much, because surprise, every data block in this dataset is like this - I found a 7M region from my low compressibility dataset which differed between the EA and non-EA runs, found that it compressed only with zstd-2 and above, and made a 42 GB dataset out of copies of that 7 MB.)

![1M evil 5900X](https://user-images.githubusercontent.com/214141/164955746-89c6c55f-e678-4694-a061-423493be7236.png)

Sadly, you can see that using zstd-2 saves you from most of the pathological compression savings behavior, but it also nukes nearly all of the performance benefit. (Not pictured above - I tried constructing custom zstd level settings to have our cake and eat it too, but unfortunately, for at least this case, the same setting both killed the speed and resulted in concluding it was compressible - the maximum window it would look back over.)

But what about something with less unlimited CPU power to burn?

**M1 (hw accelerated) VM, low compression, 128k:**
![128k lowcomp M1VM](https://user-images.githubusercontent.com/214141/164955738-daebac94-de01-48b0-8689-3f281ea3815c.png)
**M1 VM, low compression, 1M:**
![1M lowcomp M1VM](https://user-images.githubusercontent.com/214141/164955734-786e52f0-5d2e-44f4-88e3-302fd15b1142.png)
And not very interesting, but **M1 VM, high compression. 1M:**
![1M highcomp M1VM](https://user-images.githubusercontent.com/214141/164955729-70f259bb-70f6-46a0-a2e0-cd2adc5ac4e6.png)

So, worst case is visible, but not earth-shattering. (To be concrete, the largest delta between no preprocessing, EA, and alwayson was 7.9 and 63 seconds, respectively, at zstd-13/zstd-15, which was against a total baseline time of 602/846 seconds.)

Or even less CPU?

**Pi 4, low compression, 1M recordsize:**
![1M lowcomp Pi4 WIP2](https://user-images.githubusercontent.com/214141/164955726-d648f478-b0bd-4374-b21b-a77fcd1fd9c9.png)

That is, for reference, a difference between 835 seconds and 2321 seconds at zstd-15. Obviously, zstd levels above 3 are not really a gain on this particular dataset, but it’s interesting both to see how large the delta becomes, and how huge a percentage are only getting filtered out by the zstd pass here.

rincebrain commented:
There, I've dropped most of the knobs from it, cleaned it up, corrected the style complaints, squashed it to one commit, and am dropping the draft marker.

(It has #13375 baked in because crashing on a known bug doesn't do anyone any good for evaluating this.)

behlendorf commented on module/zstd/zfs_zstd.c:
```diff
@@ -135,7 +188,7 @@ static void zstd_free(void *opaque, void *ptr);
 	zstd_alloc,
 	zstd_free,
-	NULL,
+		NULL,
```
nit: stray extra tab

rincebrain replied:
I'm...not sure how that even happened, it's not like I was editing anywhere near there.

Fixed.

behlendorf commented on module/zstd/zfs_zstd.c:
```diff
@@ -62,6 +66,21 @@ typedef struct zstd_stats {
 	kstat_named_t	zstd_stat_dec_fail;
+/*
+ * LZ4 first-pass early abort verdict
+ */
```
nit: comments should be indented.

rincebrain replied:
Whoops. Fixed.

behlendorf commented on module/zstd/zfs_zstd.c:
```diff
@@ -437,8 +548,10 @@ zfs_zstd_compress(void *s_start, void *d_start, size_t s_len, size_t d_len,
+		int err = ZSTD_getErrorCode(c_len);
+		if (err != ZSTD_error_dstSize_tooSmall) {
 			ZSTDSTAT_BUMP(zstd_stat_com_fail);
+			zfs_dbgmsg("Error: %s", ZSTD_getErrorString(err));
```
How using `dprintf()` here instead.

rincebrain replied:
Sure.

behlendorf reviewed (approved):
That's for providing all the performance data and analysis.  This looks great, consider me convinced!

allanjude reviewed (approved):
Reviewed By: Allan Jude <allan@klarasystems.com>

Ornias1993 reviewed (approved):
Solid idea, great execution. 
LGTM.

gmelikov commented on module/zstd/zfs_zstd.c:
```diff
@@ -781,8 +897,8 @@ module_init(zstd_init);
-EXPORT_SYMBOL(zfs_zstd_decompress);
-EXPORT_SYMBOL(zfs_zstd_cache_reap_now);
+ZFS_MODULE_PARAM(zfs, zstd_, earlyabort_pass, INT, ZMOD_RW,
+	"Should we try early abort when using zstd?");
```
```suggestion
	"Enable early abort attempt when using zstd");
```

rincebrain replied:
I'd probably say "Enable early abort attempt**s**", but LGTM; I'll push that, thanks.

gmelikov commented on module/zstd/zfs_zstd.c:
```diff
@@ -781,8 +897,8 @@ module_init(zstd_init);
+ZFS_MODULE_PARAM(zfs, zstd_, earlyabort_pass, INT, ZMOD_RW,
+	"Should we try early abort when using zstd?");
+ZFS_MODULE_PARAM(zfs, zstd_, abort_size, UINT, ZMOD_RW,
+	"Only >= this many bytes should we try early abort");
```
```suggestion
	"Minimal size of block to attempt early abort");
```

rincebrain replied:
LGTM; will push

gmelikov reviewed (commented):
Small nits on params docstrings, what do you think about it?

rincebrain commented:
Pushed those suggestions plus remembered to add the parameters to `zfs(4)`.

Ornias1993 commented:
@BrainSlayer With this in place, I think we can move the default ZSTD algorithm (when using zstd without defining any number) from zstd-3 to zstd-5.

It would give users about the same performance, with a slight increase in compression ratio, which compensates a bit for the slight increase in size on low-comp data.

BrainSlayer commented:
@Ornias1993 the point why zstd-3 is default is just to be more comparable to gzip compression while performance is much higher. changing it to 5 will significant decrease performance. see benchmark results bellow using lzbench
`zstd 1.5.2 -1             737 MB/s  2024 MB/s    16702890  49.76 target/test.tar
zstd 1.5.2 -2             593 MB/s  1971 MB/s    15886983  47.33 target/test.tar
zstd 1.5.2 -3             420 MB/s  1925 MB/s    15020435  44.75 target/test.tar
zstd 1.5.2 -4             368 MB/s  1889 MB/s    14719456  43.85 target/test.tar
zstd 1.5.2 -5             195 MB/s  1877 MB/s    14698490  43.79 target/test.tar
zstd 1.5.2 -6             153 MB/s  1943 MB/s    14509011  43.22 target/test.tar
zstd 1.5.2 -7             140 MB/s  1944 MB/s    14195766  42.29 target/test.tar
zstd 1.5.2 -8             118 MB/s  1971 MB/s    14143771  42.14 target/test.tar
zstd 1.5.2 -9             116 MB/s  1952 MB/s    13796398  41.10 target/test.tar
zstd 1.5.2 -10           93.2 MB/s  1990 MB/s    13758540  40.99 target/test.tar
zstd 1.5.2 -11           68.3 MB/s  1979 MB/s    13744904  40.95 target/test.tar
zstd 1.5.2 -12           48.5 MB/s  1976 MB/s    13741419  40.94 target/test.tar
zstd 1.5.2 -13           38.6 MB/s  1978 MB/s    13729928  40.90 target/test.tar
zstd 1.5.2 -14           30.5 MB/s  1989 MB/s    13718800  40.87 target/test.tar
zstd 1.5.2 -15           23.9 MB/s  1996 MB/s    13698013  40.81 target/test.tar
zstd 1.5.2 -16           17.5 MB/s  1727 MB/s    13389633  39.89 target/test.tar
zstd 1.5.2 -17           12.0 MB/s  1652 MB/s    13147921  39.17 target/test.tar
zstd 1.5.2 -18           9.41 MB/s  1464 MB/s    12793148  38.11 target/test.tar
zstd 1.5.2 -19           7.66 MB/s  1465 MB/s    12768043  38.04 target/test.tar
zstd 1.5.2 -20           6.47 MB/s  1390 MB/s    12650131  37.69 target/test.tar
zstd 1.5.2 -21           5.80 MB/s  1402 MB/s    12647109  37.68 target/test.tar
zstd 1.5.2 -22           5.34 MB/s  1395 MB/s    12645416  37.67 target/test.tar
`

allanjude commented:
The default is 3, because that is zstd's default. It is not something I decided when writing the original patches.
https://github.com/facebook/zstd/blob/ef60302af962840bced324034b73d13c481946d1/lib/zstd.h#L119

Ranguvar commented on module/zstd/zfs_zstd.c:
The man page correctly notes the earlyabort heuristic runs at zstd levels >= 3 by default but the code comment for zfs_zstd_compress_wrap instead says it does not run at level 3 and is confusing
<= should be <

`* - Zeroth, if this is <= zstd-3, or < zstd_abort_size (currently 128k), don't try any of this, just go. (because experimentally that was a reasonable cutoff for a perf win with tiny ratio change)`

rincebrain replied:
Yup, just something I missed because I was dithering over which version to use. Feel free to open a tiny PR changing it, or I'll try to remember when I get a chance later.

BrainSlayer replied:
yep. in code documentation is wrong. the code itself starts heuristic at => level 3
